### PR TITLE
Fix for consistent firing of buttonup/down/touchstart/end/buttonchanged

### PR DESCRIPTION
**Pull request by DigiTec**
This change delinks these event streams so that they can fire independenty of one another. I've also added a sufficiently complete number of tests for the new behaior to make sure all events that should be firing are, looking for negative events (ensuring some events don't fire when they should not) and that the values all come through as expected.

As part of the fix I've supplied some helpers that are now consistently used in tracked-controls.test.js for verifying individual events and streams of events in a more concise way to avoid duplicating a lot of code.

I've also made consistent the use of sinon's spy asserts, which were previously using the ".called" property which reaches into the abstraction further than it should. This also allows us to upgrade certain spy verification's to look for a specific number of calls (calledOnce) which is an overall improvement.

This change cascaded into the oculus-touch-controls.test.js and gearvr-controls.test.js as well. For the oculus-touch-controls, this is mostly assert fixups, but also does some additional deep object verification.

For the gearvr-controls.test.js this included some bug fixes (using wrong event names) when I was trying to make the test fail and could not, only to realize it wasn't sniffing for the right event names. This is in preparation for creation of the oculus-go-controls component and test suite.

**Description:**

**Changes proposed:**
-
-
-

### CODE REVIEW

File: src/components/tracked-controls.js [hunk 1]
```diff
@@ -51,7 +51,9 @@ module.exports.Component = registerComponent('tracked-controls', {
   tick: function (time, delta) {
     var mesh = this.el.getObject3D('mesh');
     // Update mesh animations.
-    if (mesh && mesh.update) { mesh.update(delta / 1000); }
+    if (mesh && mesh.update) {
```
Comment (DigiTec) [inline src/components/tracked-controls.js]: Updated for consistency. It seems that the pattern for a complicated line is to break and not do it all on one line.

File: src/components/tracked-controls.js [hunk 2]
```diff
@@ -210,8 +220,8 @@ module.exports.Component = registerComponent('tracked-controls', {
    * @returns {boolean} Whether button has changed in any way.
    */
   handleButton: function (id, buttonState) {
-    var changed = this.handlePress(id, buttonState) ||
-                  this.handleTouch(id, buttonState) ||
+    var changed = this.handlePress(id, buttonState) |
```
Comment (DigiTec) [inline src/components/tracked-controls.js]: This is the core bug fix. We no longer do the short circuit operator. Instead we process all three handle* calls so that all events get generated and prior button states get saved correctly.

File: tests/components/gearvr-controls.test.js [hunk 3]
```diff
@@ -149,8 +151,8 @@ suite('gearvr-controls', function () {
       component.checkIfControllerPresent();
 
       // Fail purposely.
-      el.addEventListener('thumbstickmoved', function (evt) {
-        assert.ok(false);
+      el.addEventListener('trackpadmoved', function (evt) {
```
Comment (DigiTec) [inline tests/components/gearvr-controls.test.js]: This was previously incorrect. Likely since it was copied from the oculus-touch-controls.test.js I think. Once I changed it I was able to verify that I could indeed hit the fail when I updated the changed value accordingly.

File: tests/components/gearvr-controls.test.js [hunk 4]
```diff
@@ -206,7 +214,14 @@ suite('gearvr-controls', function () {
       var trackedControls = el.components['tracked-controls'];
       var applyArmModelSpy = this.sinon.spy(trackedControls, 'applyArmModel');
       trackedControls.tick();
-      assert.notOk(applyArmModelSpy.called);
+
+      // Verify that the function which applies arm model is not called when disabled.
+      this.sinon.assert.notCalled(applyArmModelSpy);
+
+      // Additionally verify that no other offets have been applied.
+      assert.strictEqual(el.object3D.position.x, 0);
```
Comment (DigiTec) [inline tests/components/gearvr-controls.test.js]: Added additional verification for the arm model applying.

File: tests/components/gearvr-controls.test.js [hunk 5]
```diff
@@ -216,7 +231,29 @@ suite('gearvr-controls', function () {
       var trackedControls = el.components['tracked-controls'];
       var applyArmModelSpy = this.sinon.spy(trackedControls, 'applyArmModel');
       trackedControls.tick();
-      assert.ok(applyArmModelSpy.called);
+
+      // Verify that the function which applies arm model is called.
+      this.sinon.assert.calledOnce(applyArmModelSpy);
+    });
+
+    test('verifies armModel position is applied for the right hand', function () {
```
Comment (DigiTec) [inline tests/components/gearvr-controls.test.js]: Verify right handed arm model works correctly (positive X offset)

File: tests/components/gearvr-controls.test.js [hunk 6]
```diff
@@ -216,7 +231,29 @@ suite('gearvr-controls', function () {
...
+      makePresent(el);
+      var trackedControls = el.components['tracked-controls'];
+      trackedControls.tick();
+      assert.ok(el.object3D.position.x > 0);
+    });
+
+    test('verifies armModel position is applied for the left hand', function () {
+      var el = this.el;
+      el.setAttribute('gearvr-controls', 'armModel', true);
+      el.setAttribute('gearvr-controls', 'hand', 'left');
```
Comment (DigiTec) [inline tests/components/gearvr-controls.test.js]: Verify left handed arm model works correctly (negative X offset)

File: tests/components/oculus-touch-controls.test.js [hunk 7]
```diff
@@ -28,9 +28,9 @@ suite('oculus-touch-controls', function () {
 
       component.checkIfControllerPresent();
 
-      assert.notOk(injectTrackedControlsSpy.called);
-      assert.notOk(addEventListenersSpy.called);
-      assert.ok(removeEventListenersSpy.called);
+      this.sinon.assert.notCalled(injectTrackedControlsSpy);
```
Comment (DigiTec) [inline tests/components/oculus-touch-controls.test.js]: There is a series of these updates. We shouldn't be breaking the sinon abstraction on the spy and accessing called directly if we can avoid it. Instead use the helpers for notCalled, called, calledOnce, etc...

File: tests/components/oculus-touch-controls.test.js [hunk 8]
```diff
@@ -83,10 +83,10 @@ suite('oculus-touch-controls', function () {
...
       component.checkIfControllerPresent();
 
-      assert.notOk(injectTrackedControlsSpy.called);
-      assert.notOk(addEventListenersSpy.called);
-      assert.notOk(removeEventListenersSpy.called);
-      assert.ok(component.controllerPresent);
+      this.sinon.assert.notCalled(injectTrackedControlsSpy);
+      this.sinon.assert.notCalled(addEventListenersSpy);
+      this.sinon.assert.notCalled(removeEventListenersSpy);
+      assert.strictEqual(component.controllerPresent, true);
```
Comment (DigiTec) [inline tests/components/oculus-touch-controls.test.js]: More consistently use strictEqual

File: tests/components/tracked-controls.test.js [hunk 9]
```diff
@@ -258,10 +257,7 @@ suite('tracked-controls', function () {
         call => call.args[0] === 'buttonchanged');
       assert.equal(emitChangedCalls.length, 1);
 
-      const emitCall = emitChangedCalls[0];
-      assert.equal(emitCall.args[0], 'buttonchanged');
-      assert.deepEqual(emitCall.args[1].id, 0);
-      assert.deepEqual(emitCall.args[1].state, controller.buttons[0]);
+      assertButtonEvent(emitChangedCalls[0], 'buttonchanged', 0, controller.buttons[0]);
```
Comment (DigiTec) [inline tests/components/tracked-controls.test.js]: Helper is at the bottom. Basically avoids us writing a 4 lines of assert code and consolidates into a useful single call.

File: tests/components/tracked-controls.test.js [hunk 10]
```diff
@@ -273,10 +269,129 @@ suite('tracked-controls', function () {
...
+    });
+
+    test('emits buttonchanged if value changed', function () {
```
Comment (DigiTec) [inline tests/components/tracked-controls.test.js]: New test for 'value' property.

File: tests/components/tracked-controls.test.js [hunk 11]
```diff
@@ -273,10 +269,129 @@ suite('tracked-controls', function () {
...
+      component.tick();
+
+      const emitChangedCalls = emitSpy.getCalls().filter(
+        call => call.args[0] === 'buttonchanged');
+      assert.equal(emitChangedCalls.length, 1);
+
+      assertButtonEvent(emitChangedCalls[0], 'buttonchanged', 0, controller.buttons[0]);
+    });
+
+    test('emits independent streams for buttondown, touchstart and buttonchanged', function () {
```
Comment (DigiTec) [inline tests/components/tracked-controls.test.js]: New test to verify we get separate event streams

File: tests/components/tracked-controls.test.js [hunk 12]
```diff
@@ -273,10 +269,129 @@ suite('tracked-controls', function () {
...
+      const emitChangedCalls = emitSpy.getCalls().filter(
+        call => call.args[0] === 'buttonchanged');
+      assert.equal(emitChangedCalls.length, 1);
+
+      assertButtonEvent(emitButtonCalls[0], 'buttondown', 0, controller.buttons[0]);
+      assertButtonEvent(emitTouchCalls[0], 'touchstart', 0, controller.buttons[0]);
+      assertButtonEvent(emitChangedCalls[0], 'buttonchanged', 0, controller.buttons[0]);
+    });
+
+    test('emits independent streams for buttonup, touchend and buttonchanged', function () {
```
Comment (DigiTec) [inline tests/components/tracked-controls.test.js]: Same for if we release the button, we should get all 3 event series.

File: tests/components/tracked-controls.test.js [hunk 13]
```diff
@@ -273,10 +269,129 @@ suite('tracked-controls', function () {
...
+      const emitChangedCalls = emitSpy.getCalls().filter(call => call.args[0] === 'buttonchanged');
+      assert.equal(emitChangedCalls.length, 1);
+
+      // Verify each of the 3 events has the correct event name and state.
+      assertButtonEvent(emitButtonCalls[0], 'buttonup', 0, controller.buttons[0]);
+      assertButtonEvent(emitTouchCalls[0], 'touchend', 0, controller.buttons[0]);
+      assertButtonEvent(emitChangedCalls[0], 'buttonchanged', 0, controller.buttons[0]);
+    });
+
+    test('emits correct event stream for spaced out interaction.', function () {
```
Comment (DigiTec) [inline tests/components/tracked-controls.test.js]: Verifies that multi-tick interactions produce the correct series of events for a simple press and release.

File: tests/components/tracked-controls.test.js [hunk 14]
```diff
@@ -273,10 +269,129 @@ suite('tracked-controls', function () {
...
+      // Fourth round, verify we only see a touchend and button changed when we lift our touch.
+      emitSpy = this.sinon.spy(el, 'emit');
+      controller.buttons[0].touched = false;
+      component.tick();
+
+      assertEventStream(emitSpy.getCalls(), ['buttonchanged', 'touchend'], ['buttondown', 'buttonup', 'touchstart']);
+      emitSpy.restore();
+    });
+
+    test('emits correct event states on a fast click', function () {
```
Comment (DigiTec) [inline tests/components/tracked-controls.test.js]: Verifies that if the states all change in single ticks that we still get proper event streams. Special case of the prior test that also tests that we have independent event streams.

File: tests/components/tracked-controls.test.js [hunk 15]
```diff
@@ -465,3 +568,43 @@ function toQuaternion (x, y, z) {
     return quaternion.toArray();
   })();
 }
+
+/**
```
Comment (DigiTec) [inline tests/components/tracked-controls.test.js]: New helpers so we can consolidate some existing code.

**Comment by DigiTec** [comment]
Test failure is FireFox only. Unrelated to the changes I've made unless I'm missing something. Maybe this is a known issue?

SUMMARY:
✔ 11 tests completed
✖ 1 test failed
FAILED TESTS:
  ✖ "after each" hook for "emits mousedown event on intersectedEl"
    Firefox 59.0.0 (Linux 0.0.0)
  Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
  require<["/home/travis/build/aframevr/aframe/tests/__init.test.js"]</</<@/tmp/tests/__init.test.js:57:0 <- /tmp/f633f031b631506f1dd8c8e8c60c723e.browserify:64579:5

**Comment by DigiTec** [comment]
Verified that the offending test failure is also on master (updated my local repository master to verify).

File: src/components/tracked-controls.js (hunk 1 above)
Comment (dmarcos) [reply src/components/tracked-controls.js]: Can we keep this as a one liner? In modern displays aspect ratio vertical space is at a premium.

File: src/components/tracked-controls.js [hunk 16]
```diff
@@ -156,7 +164,9 @@ module.exports.Component = registerComponent('tracked-controls', {
       object3D.position.fromArray(pose.position);
     } else {
       // Controller not 6DOF, apply arm model.
-      if (this.data.armModel) { this.applyArmModel(object3D.position); }
+      if (this.data.armModel) {
+        this.applyArmModel(object3D.position);
```
Comment (dmarcos) [inline src/components/tracked-controls.js]: Can we keep this as a one liner?

File: tests/components/gearvr-controls.test.js (hunk 3 above)
Comment (dmarcos) [reply tests/components/gearvr-controls.test.js]: good catch 👍

File: tests/components/tracked-controls.test.js [hunk 17]
```diff
@@ -465,3 +568,43 @@ function toQuaternion (x, y, z) {
...
+ * data matches.
+ *
+ * @param {object} eventCall - The spied call to emit.
+ * @param {string} eventName - The expected name of the event.
+ * @param {number} eventId - The button id firing the event.
+ * @param {object} eventState - The full event state.
+ */
+function assertButtonEvent (eventCall, eventName, eventId, eventState) {
+  assert.equal(eventCall.args[0], eventName);
+  assert.deepEqual(eventCall.args[1].id, eventId);
```
Comment (dmarcos) [inline tests/components/tracked-controls.test.js]: Should this just be `assert.equal(eventCall.args[1].id, eventId);` ?

Comment (DigiTec) [reply tests/components/tracked-controls.test.js]: yes.

File: tests/components/tracked-controls.test.js (hunk 9 above)
Comment (dmarcos) [reply tests/components/tracked-controls.test.js]: cool

File: src/components/tracked-controls.js (hunk 16 above)
Comment (DigiTec) [reply src/components/tracked-controls.js]: Sure, what is the rationale for 1 line versus line breaking? there seems to be no formatter or consistency across the code base with regards to this. The only consistency I did find was that most complex statements (like this one) are broken across lines and return clauses within the statement block tend to be on the same line.

File: src/components/tracked-controls.js (hunk 1 above)
Comment (dmarcos) [reply src/components/tracked-controls.js]: I think this is the last comment to address and this is ready to go.

**Comment by dmarcos** [comment]
@DigiTec Thanks for the fix and the test coverage. Much appreciated 👍

File: src/components/tracked-controls.js (hunk 16 above)
Comment (dmarcos) [reply src/components/tracked-controls.js]: You're right that the rule is a bit loose. If the line is short we keep it in one line. If you have a very long statement we break it in multiple lines. We don't have a specific number of characters. That's the best answer I can give 😄

**Comment by dmarcos** [comment]
Thank you!

File: tests/components/gearvr-controls.test.js [hunk 18]
```diff
@@ -204,19 +212,48 @@ suite('gearvr-controls', function () {
...
       var trackedControls = el.components['tracked-controls'];
-      var applyArmModelSpy = this.sinon.spy(trackedControls, 'applyArmModel');
+      var applyArmModelSpy = sinon.spy(trackedControls, 'applyArmModel');
       trackedControls.tick();
-      assert.notOk(applyArmModelSpy.called);
+
+      // Verify that the function which applies arm model is not called when disabled.
+      sinon.assert.notCalled(applyArmModelSpy);
+
+      // Additionally verify that no other offets have been applied.
```
Comment (vincentfretin) [inline tests/components/gearvr-controls.test.js]: offets => offsets
